### DISCUSSION TITLE
Not able to Receive Message on  IOS App

### DISCUSSION THREAD

**Discussion by freelancer-top** [invertase/react-native-firebase#7199] [Q&A]
I have followed documentation. I am able to get device token. Using that device token able to send message via Firebase . API Response says success but message on IOS device is not there . I suspect something wrong with my appdeleage.mm .  Can you please suggest what is missing ? 
---------------
#import "AppDelegate.h"
#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <Firebase.h>
#import <FirebaseMessaging/FirebaseMessaging.h>

@interface AppDelegate () <UIApplicationDelegate, FIRMessagingDelegate, UNUserNotificationCenterDelegate>
@end

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  // Configure Firebase
  [FIRApp configure];
  
  // Add Firebase messaging delegate
  [FIRMessaging messaging].delegate = self;
  // Register for push notifications
[RCTPushNotificationManager.sharedInstance registerForRemoteNotifications];
  
  self.moduleName = @"SchoolconnectForParent";
  self.initialProps = @{};

  return [super application:application didFinishLaunchingWithOptions:launchOptions];
} 

- (void)messaging:(FIRMessaging *)messaging didReceiveRegistrationToken:(NSString *)fcmToken {
    // TODO: Handle the received FCM token
    NSLog(@"Received FCM token: %@", fcmToken);
    // You can send this token to your server or use it for sending push notifications
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
#if DEBUG
  return [[NSURL alloc] initWithString:@"http://localhost:8081/index.bundle?platform=ios"];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

- (BOOL)concurrentRootEnabled
{
  return true;
}

// UserNotifications delegate methods
- (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)(void))completionHandler
{
  // Handle the notification response here
  completionHandler();
}

- (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo fetchCompletionHandler:(void (^)(UIBackgroundFetchResult))completionHandler
{
  // Handle the remote notification here
  completionHandler(UIBackgroundFetchResultNewData);
}

- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken
{
  // Forward the device token to Firebase Messaging
  [[FIRMessaging messaging] setAPNSToken:deviceToken type:FIRMessagingAPNSTokenTypeProd];
}

- (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)error
{
  NSLog(@"Failed to register for remote notifications: %@", error);
}

@end

---------

My React native  version :0.71.7 
"@react-native-firebase/app": "^18.0.0",
    "@react-native-firebase/messaging": "^18.0.0",

Here is my code to ensure Permission and token is there : ""Permisssion Granted" is shown on screen 
-------------
export const  requestUserPermission = async () => {
  const authStatus = await messaging().requestPermission();

  if (authStatus === messaging.AuthorizationStatus.AUTHORIZED) {
    Alert.alert("Permisssion Granted");
    console.log('**** Permission granted **** ');
  } else if (authStatus === messaging.AuthorizationStatus.DENIED) {
    Alert.alert("Permisssion Denied");

    console.log('Permission denied');
  }
}; 

if (messaging().isDeviceRegisteredForRemoteMessages == false )
  {
  // Register the device with FCM
  await messaging().registerDeviceForRemoteMessages();
  }

  // Get the token
  const token = await messaging().getToken();

-------------

  async function onMessageReceived(message: any) {
        Alert.alert("Yes Message Received");
        setUserMessage("Yes Received ");
        Alert.alert(message);
    }

    messaging().onMessage(onMessageReceived);
    messaging().setBackgroundMessageHandler(onMessageReceived);

**Comment by mikehardy** [comment]
Hi there. Without the notification JSON that you send via the firebase-admin SDK using the FCM APIs as a test, I cannot say why the FCM is not coming through.

There are many related issues in the repository that you may search for that have related issues.

My hunch is that you are trying to send a data-only / background notification, which are both difficult to do correctly at all (they require specific entitlements etc) and are specifically *not guaranteed* to deliver reliably. iOS will frequently just drop them, even if you do everything right.

**Reply by freelancer-top** [reply]
Here is my input JSON : {
    "data": {
        "Type": "ALERT",
        "some" : "something new here"
    },
    "priority": "high",
    "registration_ids": [
       
       "fmgZNEMTvUcGowBADpsaHs:APA91bF4vYWnKdNagEqf6yIh03ZnM_KGCJo626d3Dl1e0JWUycA-Wgij2b7yqMbWX6dW59IEJ2v7Hp1YOSGcojTYMgRH1JXBE3pTE0ski1vil8kQEARdv-Y4BJ97E78GWJA8zVQdibvG"
    ]
}


and here is output : 

{
    "multicast_id": 8448764455292048141,
    "success": 1,
    "failure": 0,
    "canonical_ids": 0,
    "results": [
        {
            "message_id": "0:1688171125043125%38657268f9fd7ecd"
        }
    ]
}

**Reply by mikehardy** [reply]
Thanks for the JSON, that helps a lot. It appears like exactly this as mentioned:

> My hunch is that you are trying to send a data-only / background notification, which are both difficult to do correctly at all (they require specific entitlements etc) and are specifically not guaranteed to deliver reliably. iOS will frequently just drop them, even if you do everything right.

This is not something you or I are in control of. Apple is in control of this. The *only* way to get *reliable* delivery on iOS is to include a `notification` block in your FCM JSON. data-only messages *may* deliver but they may not. You must not rely on them.

**Reply by toscalivia83** [reply]
@mikehardy, it seems I'm kind of having the same problem even though I can receive some notifications on ios when I open the app. I've created an issue here: https://github.com/invertase/react-native-firebase/discussions/7218
Would you suggest sending a `notifications` field within the `admin.messaging().sendAll(notifications)` of type:

```
notification: object;
data: object;
token: string;
```
So that we include both data AND notification even if they're the same?

**Reply by mikehardy** [reply]
As long as there is a notification block, your Fcm will have a much better chance of delivery. The rest is optional/ not important from a delivery perspective I think, let business needs dictate

**Reply by toscalivia83** [reply]
Thanks for your answer @mikehardy I can confirm this is working as expected with the notification block!
